drivers/serial/uart_nrfx_uarte.c

```diff
@@ -251,6 +321,468 @@ static int uarte_nrfx_config_get(struct device *dev, struct uart_config *cfg)
+	} else {
+		data->async->rx_cnt.cnt = 0;
+	}
+	data->async->rx_total_byte_cnt = 0;
+	data->async->rx_total_user_byte_cnt = 0;
+
+	data->async->rx_timeout = timeout;
+	data->async->rx_timeout_slab = timeout / RX_TIMEOUT_DIV > MS_PER_TICK ?
+				       (timeout / RX_TIMEOUT_DIV) - 1 :
+				       MS_PER_TICK;
+	data->async->rx_timeout_slab_real = data->async->rx_timeout_slab
+				      + MS_PER_TICK;
+	data->async->rx_size = len;
+	data->async->rx_buf = buf;
+	data->async->rx_offset = 0;
+	nrf_uarte_rx_buffer_set(cfg->uarte_regs, buf, len);
+
+	nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_ENDRX);
+	nrf_uarte_event_clear(cfg->uarte_regs, NRF_UARTE_EVENT_RXSTARTED);
+
+	nrf_uarte_task_trigger(cfg->uarte_regs, NRF_UARTE_TASK_STARTRX);
+	data->async->rx_enabled = true;
+	return 0;
+}
+
+static int uarte_nrfx_rx_buf_rsp(struct device *dev, u8_t *buf, size_t len)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	const struct uarte_nrfx_config *cfg = get_dev_config(dev);
+
+	if (data->async->rx_next_buf == NULL) {
+		data->async->rx_next_buf = buf;
+		nrf_uarte_rx_buffer_set(cfg->uarte_regs, buf, len);
+	} else {
+		return -EBUSY;
+	}
+	return 0;
+}
+
+static int uarte_nrfx_callback_set(struct device *dev, uart_callback_t callback,
+				  void *user_data)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	data->async->evt_handler = callback;
+	data->async->user_data = user_data;
+
+	return 0;
+}
+
+static int uarte_nrfx_rx_disable(struct device *dev)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	const struct uarte_nrfx_config *cfg = get_dev_config(dev);
+
+	if (data->async->rx_buf == NULL) {
+		return -EFAULT;
+	}
+	k_delayed_work_cancel(&data->async->rx_timeout_work);
+	data->async->rx_enabled = false;
+
+	nrf_uarte_task_trigger(cfg->uarte_regs, NRF_UARTE_TASK_STOPRX);
+
+	return 0;
+}
+
+static void tx_timeout(struct device *dev)
+{
+	(void) uarte_nrfx_tx_abort(dev);
+}
+
+static void user_callback(struct device *dev, struct uart_event *evt)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+
+	if (data->async->evt_handler) {
+		data->async->evt_handler(evt, data->async->user_data);
+	}
+}
+
+/**
+ * Whole timeout is divided by RX_TIMEOUT_DIV into smaller units, rx_timeout
+ * is executed periodically every rx_timeout_slab ms. If between executions
+ * data was received, then we start counting down time from start, if not, then
+ * we subtract rx_timeout_slab from rx_timeout_left.
+ * If rx_timeout_left is less than rx_timeout_slab it means that receiving has
+ * timed out and we should tell user about that.
+ */
+static void rx_timeout(struct device *dev)
+{
+	struct uarte_nrfx_data *data = get_dev_data(dev);
+	u32_t read;
+
+	if (hw_rx_counting_enabled(data)) {
+		read = nrfx_timer_capture(&data->async->timer, 0);
+	} else {
+		read = data->async->rx_cnt.cnt;
+	}
+
+	/* Check if data was received since last function call */
+	if (read > data->async->rx_total_byte_cnt) {
```

Comment:
Wouldn't it be safer to use `!=`? In theory, when a continuous receiving is enabled with next buffers provided consecutively, the total counter could wrap around.